### PR TITLE
add functions for stepper PIN enabling and disabling

### PR DESCRIPTION
**Pull request by joshua255**
making a pull request for @sambosch77's work, not ready to merge yet.

### CODE REVIEW

File: ursa.ino [hunk 1]
```diff
@@ -146,7 +146,13 @@ void loop() {  // on core 1. the balencing control loop will be here, with the g
     digitalWrite(LED_BUILTIN, HIGH);
 
     if (!wasRobotEnabled) {  // the robot wasn't enabled, but now it is, so this must be the first loop since it was enabled. re set up anything you might want to
-      // TODO: turn on stepper motors
+      // TODO: turn on stepper motors (complete?)
+      void setupMotors() { // function turns on stepper motors using an enable PIN on the drivers
+      pinMode(ENS_PIN GPIO_NUM_23, OUTPUT) // sets this PIN as an output
+      }
+      void loop() {
```
Comment (joshua255) [inline ursa.ino]: a digitalWrite changes the pin and it won't change again until there is a new digital write. this void loop is uncecessary and is probably why the code won't compile

File: ursa.ino [hunk 2]
```diff
@@ -146,7 +146,13 @@ void loop() {  // on core 1. the balencing control loop will be here, with the g
     digitalWrite(LED_BUILTIN, HIGH);
 
     if (!wasRobotEnabled) {  // the robot wasn't enabled, but now it is, so this must be the first loop since it was enabled. re set up anything you might want to
-      // TODO: turn on stepper motors
+      // TODO: turn on stepper motors (complete?)
+      void setupMotors() { // function turns on stepper motors using an enable PIN on the drivers
+      pinMode(ENS_PIN GPIO_NUM_23, OUTPUT) // sets this PIN as an output
```
Comment (joshua255) [inline ursa.ino]: pinMode should be put in setup, and only needs to be called once

File: ursa.ino [hunk 3]
```diff
@@ -146,7 +146,13 @@ void loop() {  // on core 1. the balencing control loop will be here, with the g
     digitalWrite(LED_BUILTIN, HIGH);
 
     if (!wasRobotEnabled) {  // the robot wasn't enabled, but now it is, so this must be the first loop since it was enabled. re set up anything you might want to
-      // TODO: turn on stepper motors
+      // TODO: turn on stepper motors (complete?)
+      void setupMotors() { // function turns on stepper motors using an enable PIN on the drivers
```
Comment (joshua255) [inline ursa.ino]: you are declaring the function void setupMotors inside the main loop, you can either delete the function and leave only the digitalWrite, or declare the function (with just a digital write inside) outside any other functions and then call it here. Also, using the name enableMotors would be more descriptive, keeping the name setupMotors if you want a function to set the pinmode to output in setup.

File: ursa.ino [hunk 4]
```diff
@@ -146,7 +146,13 @@ void loop() {  // on core 1. the balencing control loop will be here, with the g
...
+      pinMode(ENS_PIN GPIO_NUM_23, OUTPUT) // sets this PIN as an output
+      }
+      void loop() {
+      digitalWrite(ENS_PIN GPIO_NUM_23, HIGH) // loop in place to make sure PIN stays enabled (not sure if necessary)
```
Comment (joshua255) [inline ursa.ino]: needs a semicolon at the end of this line (you're not in Kotlin anymore) :smile:

File: ursa.ino [hunk 5]
```diff
@@ -178,7 +184,13 @@ void loop() {  // on core 1. the balencing control loop will be here, with the g
     timerAlarmWrite(rightStepTimer, 10000000000000000, true);  // 1Mhz / # =  rate
     leftMotorSpeed = 0;
     rightMotorSpeed = 0;
-    // TODO: turn off stepper motors
+    // TODO: turn off stepper motors (complete?)
+    void turnOffMotors (){
+    pinMode(ENS_PIN GPIO_NUM_23, OUTPUT) // making sure that this is still recognized as an output (not sure if necessary)
```
Comment (joshua255) [inline ursa.ino]: like before, pinmode is unnecessary, it can be setup in void setup and then stays forever

File: ursa.ino [hunk 6]
```diff
@@ -178,7 +184,13 @@ void loop() {  // on core 1. the balencing control loop will be here, with the g
     timerAlarmWrite(rightStepTimer, 10000000000000000, true);  // 1Mhz / # =  rate
     leftMotorSpeed = 0;
     rightMotorSpeed = 0;
-    // TODO: turn off stepper motors
+    // TODO: turn off stepper motors (complete?)
+    void turnOffMotors (){
```
Comment (joshua255) [inline ursa.ino]: you don't declare functions inside your other code, I would recommend not trying to make a function and just leaving a digital write here

File: ursa.ino [hunk 7]
```diff
@@ -178,7 +184,13 @@ void loop() {  // on core 1. the balencing control loop will be here, with the g
...
+    pinMode(ENS_PIN GPIO_NUM_23, OUTPUT) // making sure that this is still recognized as an output (not sure if necessary)
+    }
+    void loop() {
+    digitalWrite(ENS_PIN GPIO_NUM_23, LOW) // same with the turning on - making sure the steppers remain disabled
```
Comment (joshua255) [inline ursa.ino]: counter intuitively, the motor driver disables when the pin is pulled HIGH, not low.
you also need a semicolon

File: ursa.ino [hunk 8]
```diff
@@ -178,7 +184,13 @@ void loop() {  // on core 1. the balencing control loop will be here, with the g
     timerAlarmWrite(rightStepTimer, 10000000000000000, true);  // 1Mhz / # =  rate
     leftMotorSpeed = 0;
     rightMotorSpeed = 0;
-    // TODO: turn off stepper motors
+    // TODO: turn off stepper motors (complete?)
+    void turnOffMotors (){
+    pinMode(ENS_PIN GPIO_NUM_23, OUTPUT) // making sure that this is still recognized as an output (not sure if necessary)
+    }
+    void loop() {
```
Comment (joshua255) [inline ursa.ino]: like before, no loop is necessary

File: ursa.ino (hunk 4 above)
Comment (joshua255) [inline ursa.ino]: From a4988 datasheet: Enable Input - This input turns on or off all of the FET outputs. When set to a logic high, the outputs are disabled. When set to a logic low, the internal control enables the outputs as required.

File: ursa.ino (hunk 2 above)
Comment (joshua255) [inline ursa.ino]: One more thing, use just `ENS_PIN` to refer to the pin, it equals GPIO_NUM_23 which equals the number pin for the esp32 to change

File: ursa.ino [hunk 9]
```diff
@@ -146,7 +147,8 @@ void loop() {  // on core 1. the balencing control loop will be here, with the g
     digitalWrite(LED_BUILTIN, HIGH);
 
     if (!wasRobotEnabled) {  // the robot wasn't enabled, but now it is, so this must be the first loop since it was enabled. re set up anything you might want to
-      // TODO: turn on stepper motors
+      // TODO: turn on stepper motors (complete?)
```
Comment (andrewda) [inline ursa.ino]:
```suggestion
```

File: ursa.ino [hunk 10]
```diff
@@ -178,7 +180,8 @@ void loop() {  // on core 1. the balencing control loop will be here, with the g
     timerAlarmWrite(rightStepTimer, 10000000000000000, true);  // 1Mhz / # =  rate
     leftMotorSpeed = 0;
     rightMotorSpeed = 0;
-    // TODO: turn off stepper motors
+    // TODO: turn off stepper motors (complete?)
```
Comment (andrewda) [inline ursa.ino]:
```suggestion
```

**Review by andrewda — CHANGES_REQUESTED** [review]

**Comment by joshua255** [comment]
@SamBosh77 looks great! Just take Andrew's suggestion of removing comments and I think your're ready to merge!

**Review by andrewda — APPROVED** [review]
LGTM

**Pull request merged**